interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,273 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.testing.integration;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.DelayQueue;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+public final class TrafficControlProxy {
+
+  private static final int DEFAULT_BAND_BPS = 1024 * 1024;
+  private static final int DEFAULT_DELAY_NANOS = 200 * 1000 * 1000;
+  private static final Logger logger = Logger.getLogger(TrafficControlProxy.class.getName());
+
+  // TODO: make ports arguments (and probably loopback/ip as well)
+  private String loopback = "127.0.0.1";
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int queueLength;
+  private int chunkSize;
+  private int bandwidth;
+  private long latency;
+  private volatile boolean shutDown;
+  private ServerSocket clientAcceptor;
+  private Socket serverSock;
+  private Socket clientSock;
+  private List<Thread> workers = new ArrayList<Thread>();
+
+  /**
+   * Returns a new TrafficControlProxy with default bandwidth and latency.
+   */
+  public TrafficControlProxy() {
+    this(DEFAULT_BAND_BPS, DEFAULT_DELAY_NANOS, TimeUnit.NANOSECONDS);
+  }
+
+  /**
+   * Returns a new TrafficControlProxy with bandwidth set to targetBPS, and latency set to
+   * targetLatency in latencyUnits.
+   */
+  public TrafficControlProxy(int targetBps, int targetLatency, TimeUnit latencyUnits) {
+    checkArgument(targetBps > 0);
+    checkArgument(targetLatency > 0);
+    bandwidth = targetBps;
+    // divide by 2 because latency is applied in both directions
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+  }
+
+  /**
+   * Convenience method to change bandwidth/latency without restarting.
+   */
+  public void reset(int targetBps, int targetLatency, TimeUnit latencyUnits) throws IOException {
+    checkArgument(targetBps > 0);
+    checkArgument(targetLatency > 0);
+    shutDown();
+    bandwidth = targetBps;
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+    workers.clear();
```

Comment:
Maybe clear the workers in `shutDown()`, since it seems you allow shutting down and then starting again?